Sources/StreamChat/Workers/MessageUpdater.swift

```diff
@@ -277,8 +319,36 @@ class MessageUpdater: Worker {
         messageId: MessageId,
         completion: ((Error?) -> Void)? = nil
     ) {
-        apiClient.request(endpoint: .deleteReaction(type, messageId: messageId)) {
-            completion?($0.error)
+        var reaction: MessageReactionDTO?
+
+        database.write { session in
+            do {
+                reaction = try session.removeReaction(from: messageId, type: type)
+                completion?(nil)
+            } catch {
+                log.error("Failed to remove the reaction from to the database: \(error)")
+                completion?(error)
+            }
+            
+            // if session.removeReaction failed we fire the API call, worst case the call with have no effect
+            guard let reaction = reaction else {
+                self.apiClient.request(endpoint: .deleteReaction(type, messageId: messageId)) { _ in }
+                return
+            }
+
+            reaction.localState = .deleting
+
+            // TODO: verify that this is OK and that does not cause problems (api call inside db write)
+            self.apiClient.request(endpoint: .deleteReaction(type, messageId: messageId)) { result in
+                guard let error = result.error else {
+                    return
+                }
+
+                /// update the localState depending on the kind of API error
+                self.database.write { _ in
+                    reaction.localState = ClientError.isEphemeral(error: error) ? .pendingDelete : .deletingFailed
+                }
+            }
```

Comment:
The `database.write` has a completion block to notify when the database writes finishes:
```swift
database.write { session
    try session.removeReaction(from: messageId, type: type)
} completion: { error in
   if let error = error {
       ....
       return
   }
   ...
}
```

Either way, I don't think there's any issue with the current approach 👍